### DISCUSSION TITLE
Krux on Maix Cube K210

### DISCUSSION THREAD

**Discussion by Tommiller0515** [selfcustody/krux#316] [General]
Has anyone looked into possibility of running krux on a Maix Cube K210?

**Comment by odudex** [comment]
I set up a custom firmware for one once, but I don't have one with me to maintain it. The issue with Cube is its price is very close to Amigo's, and Amigo has many advantages over Cube. Amigo's larger touch screen has a much better UX.
Do you already have a Cube with you?

**Comment by Tommiller0515** [comment]
I agree, the Amigo's are great. I don't have a cube and have never tried one but thought it might be similar to UX of the DIY ones but with a battery. The lowest price I could find the Amigo for on AliExpress now is $57 and $36 for the Cube. Both have higher prices but those were the lowest I could find there.  Anyway, nice to have options and alternatives.

**Comment by odudex** [comment]
Last time I checked it was close to $50. For $36 it starts to get interesting!

**Comment by odudex** [comment]
Ok, I bought one. Let's see how it goes.

**Reply by odudex** [reply]
Hi Sergio,
My Cube is on the way, and I'm confident it will be possible to port Krux to it.
And the good news is Krux was already ported to your second device, Yaboom AIMOTION, it is available to test on [beta](https://github.com/odudex/krux_binaries), and will be available on stable release soon.
![image](https://github.com/selfcustody/krux/assets/5049478/0e11a612-20bb-403c-8a85-bddd4ab461a9)

**Comment by Tommiller0515** [comment]
Looking forward to hearing the results. Hopefully Chinese holiday doesn't slow down delivery too much

**Reply by odudex** [reply]
Firmware for Cube available to test here:
https://github.com/odudex/krux_binaries

![image](https://github.com/selfcustody/krux/assets/5049478/ac55348c-3670-4d14-905c-d31f01f91fd8)

**Reply by Tommiller0515** [reply]
That looks great! Unfortunately I tried flashing to the cube I recently received and while it appeared to be flashing it,  it never booted up - stayed on the welcome screen. When I power-cycled it, it just stayed blank. I flashed the factory default firmware from Speed wiki and it loaded back to the Welcome to MaixPy screen, so I think the cube I have is OK.

**Reply by odudex** [reply]
Sorry to hear that. I'll try to find other devices to experiment with. In Amigo we had hardware variations that needed to be addressed, it's not the case for Cube.
Some things you could try:
- Please make sure you flashed Cube's firmware, not Amigo's or Bit's. (They share the same USB ID, so that would be possible)
`./ktool-linux -B goE -b 1500000 maixpy_cube/kboot.kfpkg`
- Debug serial output, it has two ports, you must try both, press reset button while connected and check boot message. You could use Screen on Linux `screen /dev/ttyUSB0 115200` or `screen /dev/ttyUSB1 115200`. Or Putty if you are on Windows (try 2 COMx ports at 115200 baud rate)
- You could also try a full flash erase before flashing Krux firmware:
`./ktool-linux -B goE -b 1500000 -E`

**Reply by Tommiller0515** [reply]
Tried erasing and flashing. Ends with rebooting... but cube doesn't reboot. Power-cycled cube and screen just blank. Here's printout of erase and flash:

C:\Krux\krux_binaries>ktool-win.exe -B goE -b 1500000 -E
[INFO] COM Port Auto Detected, Selected  COM4
[INFO] Default baudrate is 115200 , later it may be changed to the value you set.
[INFO] Trying to Enter K210 ROM ISP Mode...
*
[INFO] ROM ISP detected, loading 2nd stage ISP
[INFO] ISP loaded in 0.914s
[INFO] Starting 2nd stage ISP at 0x805e0000
[INFO] Wait For 0.1 second for ISP to Boot
[INFO] 2nd stage ISP ok
[INFO] Selected Baudrate: 1500000
[INFO] Baudrate changed, greeting with ISP again ...
[INFO] 2nd stage ISP ok
[INFO] Initialize K210 SPI Flash
[INFO] Flash initialized successfully
[INFO] Flash ID: 0xEF6018, unique ID: EB00005104400220, size: 16 MB
[INFO] Erasing the whole SPI Flash.
[INFO] SPI Flash erased.

C:\Krux\krux_binaries>ktool-win.exe -B goE -b 1500000 maixpy_cube\kboot.kfpkg
[INFO] COM Port Auto Detected, Selected  COM4
[INFO] Default baudrate is 115200 , later it may be changed to the value you set.
[INFO] Trying to Enter K210 ROM ISP Mode...
*
[INFO] ROM ISP detected, loading 2nd stage ISP
[INFO] ISP loaded in 0.923s
[INFO] Starting 2nd stage ISP at 0x805e0000
[INFO] Wait For 0.1 second for ISP to Boot
[INFO] 2nd stage ISP ok
[INFO] Selected Baudrate: 1500000
[INFO] Baudrate changed, greeting with ISP again ...
[INFO] 2nd stage ISP ok
[INFO] Initialize K210 SPI Flash
[INFO] Flash initialized successfully
[INFO] Flash ID: 0xEF6018, unique ID: EB00005104400220, size: 16 MB
[INFO] Flashing firmware block with SHA suffix
Programming BIN: |=====================================================================================| 100.0% 77kiB/s
[INFO] Flashed 147704 B [3 chunks of 65536B] (00000000~0002FFFF) in 2.490s
[INFO] Rebooting...


C:\Krux\krux_binaries>

**Reply by tadeubas** [reply]
Can you install [MaixPy IDE](https://dl.sipeed.com/shareURL/MAIX/MaixPy/ide/v0.2.5) to help debugging?

`Tools > Open Terminal > New Terminal > Connect to serial port > Select a COM port available` (if didn't work, try another COM port)

It will show the terminal and some messages

![image](https://github.com/selfcustody/krux/assets/5518496/2a5382ee-555d-4038-aafb-0ad231820321)

![image](https://github.com/selfcustody/krux/assets/5518496/a34a0bf8-5ca7-4e59-8dfc-e2dfc9336de5)

**Comment by Tommiller0515** [comment]
![DefaultFirmware](https://github.com/selfcustody/krux/assets/158013490/f02cd6e6-b5f7-4f64-93d5-e53ba5fd782d)
![MaixCubeKrux](https://github.com/selfcustody/krux/assets/158013490/22fd9b3f-1c97-4ce6-89b5-aa670f9f45b0)

Default firmware and krux firmware for Maix Cube.

**Reply by tadeubas** [reply]
lol 🤣 :
"fuck the chip is dead!"
"failed with exit code 666"

Maybe you should find your vendor and tell about these messages? 🤣

Let's wait if @odudex has any insight into what's going on...

**Reply by tadeubas** [reply]
The image shows all the log available in the terminal?

**Reply by Tommiller0515** [reply]
Yes, but it also has the No Image area on the right side too. After erasing and flashing default firmware, the image became upside down, offset to one side, and blue background. It was originally orange background, right side up, and slightly offset to one side. My device may have issues so just in case, I'm looking into refund/replacement.

**Reply by tadeubas** [reply]
It has battery, you could try to erase everything again with `ktool-win.exe -B goE -b 1500000 -E` and left the device on until it dies. Then try again to flash the default Maixpy firmware to see what happens to the image.

**Reply by Tommiller0515** [reply]
Actually it never has stayed on when unplugged no matter how long it's been connected, so one more reason to look into return/replacement.

**Reply by tadeubas** [reply]
Yes, it has an internal 200mAh lithium battery, specs here: https://wiki.sipeed.com/soft/maixpy/en/develop_kit_board/maix_cube.html

**Reply by odudex** [reply]
It looks like some kind of data corruption to me.
Try downloading binaries again.
You can also test to flash the firmware.bin instead of boot.kfpkg
`ktool-win.exe -B goE -b 1500000 maixpy_cube\firmware.bin`

If that doesn't work, maybe there's a physical issue on flash or SoC

**Reply by Tommiller0515** [reply]
Thanks for all the help. My fault due to my github blunder. I forgot github is different in that downloading requires clicking through on the file name and then downloading raw file. Instead, I was erroneously right clicking on the file name and using save-as to download the file. It downloads and appears to work that way but must be corrupting the data somehow. I scanned a seed QR and all working very nice. My device battery isn't charging but that's a different issue. Thanks again and sorry for bothering you

**Reply by odudex** [reply]
Great news! Happy it helped.
And did you flash your Yahboom too?
About battery, what does the battery indicator shows on the screen? It may have issues when the host is USC-C, if you used one,  maybe try an old style USB-A charger.

**Reply by tadeubas** [reply]
Battery issues:

#### USB-A:
Your device should charge and turn on when connected to a USB-A port, even if it was initially turned off. You can also turn off the device while it continues to charge. However, some hosts' USB-A ports may behave like USB-C ports, as described below.

#### USB-C:

- If the device is turned off and connected to a USB-C port, it should turn on and start charging. You can turn it off again, and it will continue to charge.

- If the device is already turned on and connected to a USB-C port, it may not charge or be recognized by the computer. In this case, turn off the device to initiate recognition and charging. Once turned off and reconnected, the device should restart, be recognized by the computer, and charging should be triggered by USB-C hosts.
If your device is not charging or being recognized as expected, try using a different USB port or a different computer to determine if the issue is with the device or the host's USB port.

**Reply by Tommiller0515** [reply]
I have tried various cables and sources but battery has never charged. Krux cube display sometimes shows a green outline of battery - like it's empty, not filled in solid and not changing.

**Reply by Tommiller0515** [reply]
I don't have a Yahboom. I believe that was another person - Sergio?

**Reply by odudex** [reply]
Ops, sorry 🙃

**Reply by tadeubas** [reply]
So you have a problem with your device's battery. And did Krux work well? Did it read QR codes and display information on the screen correctly?

**Reply by Tommiller0515** [reply]
Yes. It works fine as long as it's plugged in. Reads QR codes perfectly fine, displays correctly on screen. Will this Cube work with thermal printer like Amigo. Schematic doesn't show rx/tx pins assigned to grove connector.

**Reply by tadeubas** [reply]
Yes, it has the Grove connector as shown on the photo below. You will need to connect ground and a pin to use as TX in Krux and as RX on the printer, the TX pin is configurable in Krux settings, more info here: https://github.com/selfcustody/krux/discussions/312

![maixcube_resources](https://github.com/selfcustody/krux/assets/5518496/b134101a-8f42-4475-adf6-7ba205ac6f1f)

**Reply by Tommiller0515** [reply]
OK. It looks like grove pin 2 is IO 25 for the cube and IO 7 for Amigo. That being the case, would need to set thermal printer tx to 25 for the cube. That link you provided was actually you helping me. It worked like a charm on the Amigo. Thank you! However, the way the cube case is designed covering all the way around the grove connector, it blocks mating with the grove plug. The Amigo doesn't have that problem, so I'll probably just use it. Thanks again!

**Comment by jdlcdl** [comment]
> Reason: fuck the chip is dead!
> interesting, something's wrong, boot failed with exit code 666, go to find your vendor.

Leaving a comment here so I can find this again.  Above is an interesting console error message for a K210 MAIX CUBE BRICK.
I don't remember seeing the first-32-registers dumped in the past.

I do see a similar on Amigo after a full-chip erase via ktool:
"interesting, something's wrong, boot failed with exit code 233, go to find your vendor."

It's great that this is resolved for yet another device.
